lsp/nls/src/requests/completion.rs

```diff
@@ -21,42 +29,194 @@ pub fn handle_completion(
+    // idealy we should get the id and use it instead of comparing strings labels
+    let lin_env = &linearization.lin_env;
 
-    Trace::enrich(&id, linearization);
+    // Try your best to find all record fields
+    fn find_record_fields(
+        linearization: &Vec<LinearizationItem<Types>>,
+        id: usize,
+    ) -> Option<Vec<Ident>> {
+        linearization.iter().find_map(|item| match item.kind {
+            TermKind::Record(ref fields) if item.id == id => {
+                Some(fields.keys().map(|ident| ident.clone()).collect())
+            }
+            TermKind::Declaration(_, _, ValueState::Known(new_id))
+            | TermKind::Usage(UsageState::Resolved(new_id))
+                if item.id == id =>
+            {
+                find_record_fields(linearization, new_id)
+            }
+            _ => None,
+        })
+    }
 
-    let item = linearization.item_at(&locator);
+    fn find_record_contract(
+        linearization: &Vec<LinearizationItem<Types>>,
+        id: usize,
+    ) -> Option<Contract> {
+        linearization.iter().find_map(|item| match &item.meta {
+            Some(MetaValue { contracts, .. }) if item.id == id => {
+                contracts.iter().find_map(|contract| {
+                    if let Types(AbsType::Record(..)) = contract.types {
+                        Some(contract.clone())
+                    } else {
+                        None
+                    }
+                })
+            }
+            _ if item.id == id => match item.kind {
+                TermKind::Declaration(_, _, ValueState::Known(new_id))
+                | TermKind::Usage(UsageState::Resolved(new_id)) => {
+                    find_record_contract(&linearization, new_id)
+                }
+                _ => None,
+            },
+            _ => None,
+        })
+    }
 
+    /// Extract identifiers from a row type which forms a record.
+    /// Ensure that ty is really a row type.
+    fn extract_ident(ty: &Box<Types>) -> Vec<Ident> {
+        ty.iter_as_rows()
+            .filter_map(|item| match item {
+                RowIteratorItem::Row(ident, _) => Some(ident.clone()),
+                RowIteratorItem::Tail(..) => None,
+            })
+            .collect::<Vec<_>>()
+    }
+
+    let item = linearization.item_at(&locator);
     if item == None {
+        // NOTE: this might result in a case where completion is not working.
+        // This happens when an item cannot be found from the linearization.
         server.reply(Response::new_ok(id, Value::Null));
         return Ok(());
     }
-
     let item = item.unwrap().to_owned();
 
-    let in_scope: Vec<_> = linearization
-        .get_in_scope(&item)
+    Trace::enrich(&id, linearization);
+
+    let trigger = params.context.and_then(|context| context.trigger_character);
+    let in_scope = match trigger {
+        // Record Completion
+        Some(s) if s == "." => {
```

Comment:
Silly question from a total Rust noob: can't we do this instead?
```suggestion
        Some(".") => {
```

Reply:
Yes we can do this, but we need to map `trigger` with a the transformation: `String -> &str `, because string literals have the type `&str`, but the LSP is giving us a `String`